horovod/mxnet/mpi_ops.cc

```diff
@@ -277,49 +307,19 @@ extern "C" int horovod_mxnet_broadcast_async(NDArray* input, NDArray* output,
                                              int priority) {
   MX_API_BEGIN();
 
-  std::string op_name = GetOpName("broadcast", name);
-
 #if HAVE_CUDA && !HOROVOD_GPU_BROADCAST
-  // Make async copy of input tensor to CPU tensor.
-  auto hvd_cpu_buffer = std::make_shared<MXTemporaryBuffer<NDArray>>(
-      CPU_DEVICE_ID, input->dtype());
-  auto cpu_tensor = hvd_cpu_buffer->tensor();
-  TensorUtil::AsyncCopyCudaToCPU(input, cpu_tensor);
-
-  // In-place
-  auto broadcast_async_cpu_fn = [hvd_cpu_buffer, op_name,
-                                 root_rank](RunContext rctx,
-                                            CallbackOnComplete on_complete) mutable {
-    DoBroadcastCudaOnCPU(hvd_cpu_buffer, root_rank, op_name, on_complete);
-  };
-
-  Engine::Get()->PushAsync(broadcast_async_cpu_fn, EXEC_CTX,
-                           {}, {cpu_tensor->var()},
-                           FnProperty::kCPUPrioritized, priority,
-                           "HorovodBroadcast");
-
-  // Make async copy of CPU tensor to output tensor.
-  TensorUtil::AsyncCopyCPUToCuda(cpu_tensor, output);
-#else
-  auto broadcast_async_fn = [input, output, op_name,
-                             root_rank](RunContext rctx,
-                                        CallbackOnComplete on_complete) mutable {
-    DoBroadcast(input, output, root_rank, op_name, on_complete);
-  };
+  if (input->ctx().dev_mask() == cpu::kDevMask &&
```

Comment:
Seems like this logic is now duplicated and could be abstracted away as well?

Reply:
This is actually the check I wanted to add here. If both input and output are in CPU context, there is no need to perform copy from and to CPU.

Reply:
Nevermind, found few differences between blocks upon closer inspection.